<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Anna Plantinga, Nehemiah Wilson, Haotian Zheng, Xiang Zhan, Michael Wu, Jun Chen, Ni Zhao" />


<title>MiRKAT Package Vignette</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<script>$(document).ready(function(){
    if (typeof $('[data-toggle="tooltip"]').tooltip === 'function') {
        $('[data-toggle="tooltip"]').tooltip();
    }
    if ($('[data-toggle="popover"]').popover === 'function') {
        $('[data-toggle="popover"]').popover();
    }
});
</script>
<style type="text/css">
.lightable-minimal {
border-collapse: separate;
border-spacing: 16px 1px;
width: 100%;
margin-bottom: 10px;
}
.lightable-minimal td {
margin-left: 5px;
margin-right: 5px;
}
.lightable-minimal th {
margin-left: 5px;
margin-right: 5px;
}
.lightable-minimal thead tr:last-child th {
border-bottom: 2px solid #00000050;
empty-cells: hide;
}
.lightable-minimal tbody tr:first-child td {
padding-top: 0.5em;
}
.lightable-minimal.lightable-hover tbody tr:hover {
background-color: #f5f5f5;
}
.lightable-minimal.lightable-striped tbody tr:nth-child(even) {
background-color: #f5f5f5;
}
.lightable-classic {
border-top: 0.16em solid #111111;
border-bottom: 0.16em solid #111111;
width: 100%;
margin-bottom: 10px;
margin: 10px 5px;
}
.lightable-classic tfoot tr td {
border: 0;
}
.lightable-classic tfoot tr:first-child td {
border-top: 0.14em solid #111111;
}
.lightable-classic caption {
color: #222222;
}
.lightable-classic td {
padding-left: 5px;
padding-right: 5px;
color: #222222;
}
.lightable-classic th {
padding-left: 5px;
padding-right: 5px;
font-weight: normal;
color: #222222;
}
.lightable-classic thead tr:last-child th {
border-bottom: 0.10em solid #111111;
}
.lightable-classic.lightable-hover tbody tr:hover {
background-color: #F9EEC1;
}
.lightable-classic.lightable-striped tbody tr:nth-child(even) {
background-color: #f5f5f5;
}
.lightable-classic-2 {
border-top: 3px double #111111;
border-bottom: 3px double #111111;
width: 100%;
margin-bottom: 10px;
}
.lightable-classic-2 tfoot tr td {
border: 0;
}
.lightable-classic-2 tfoot tr:first-child td {
border-top: 3px double #111111;
}
.lightable-classic-2 caption {
color: #222222;
}
.lightable-classic-2 td {
padding-left: 5px;
padding-right: 5px;
color: #222222;
}
.lightable-classic-2 th {
padding-left: 5px;
padding-right: 5px;
font-weight: normal;
color: #222222;
}
.lightable-classic-2 tbody tr:last-child td {
border-bottom: 3px double #111111;
}
.lightable-classic-2 thead tr:last-child th {
border-bottom: 1px solid #111111;
}
.lightable-classic-2.lightable-hover tbody tr:hover {
background-color: #F9EEC1;
}
.lightable-classic-2.lightable-striped tbody tr:nth-child(even) {
background-color: #f5f5f5;
}
.lightable-material {
min-width: 100%;
white-space: nowrap;
table-layout: fixed;
font-family: Roboto, sans-serif;
border: 1px solid #EEE;
border-collapse: collapse;
margin-bottom: 10px;
}
.lightable-material tfoot tr td {
border: 0;
}
.lightable-material tfoot tr:first-child td {
border-top: 1px solid #EEE;
}
.lightable-material th {
height: 56px;
padding-left: 16px;
padding-right: 16px;
}
.lightable-material td {
height: 52px;
padding-left: 16px;
padding-right: 16px;
border-top: 1px solid #eeeeee;
}
.lightable-material.lightable-hover tbody tr:hover {
background-color: #f5f5f5;
}
.lightable-material.lightable-striped tbody tr:nth-child(even) {
background-color: #f5f5f5;
}
.lightable-material.lightable-striped tbody td {
border: 0;
}
.lightable-material.lightable-striped thead tr:last-child th {
border-bottom: 1px solid #ddd;
}
.lightable-material-dark {
min-width: 100%;
white-space: nowrap;
table-layout: fixed;
font-family: Roboto, sans-serif;
border: 1px solid #FFFFFF12;
border-collapse: collapse;
margin-bottom: 10px;
background-color: #363640;
}
.lightable-material-dark tfoot tr td {
border: 0;
}
.lightable-material-dark tfoot tr:first-child td {
border-top: 1px solid #FFFFFF12;
}
.lightable-material-dark th {
height: 56px;
padding-left: 16px;
padding-right: 16px;
color: #FFFFFF60;
}
.lightable-material-dark td {
height: 52px;
padding-left: 16px;
padding-right: 16px;
color: #FFFFFF;
border-top: 1px solid #FFFFFF12;
}
.lightable-material-dark.lightable-hover tbody tr:hover {
background-color: #FFFFFF12;
}
.lightable-material-dark.lightable-striped tbody tr:nth-child(even) {
background-color: #FFFFFF12;
}
.lightable-material-dark.lightable-striped tbody td {
border: 0;
}
.lightable-material-dark.lightable-striped thead tr:last-child th {
border-bottom: 1px solid #FFFFFF12;
}
.lightable-paper {
width: 100%;
margin-bottom: 10px;
color: #444;
}
.lightable-paper tfoot tr td {
border: 0;
}
.lightable-paper tfoot tr:first-child td {
border-top: 1px solid #00000020;
}
.lightable-paper thead tr:last-child th {
color: #666;
vertical-align: bottom;
border-bottom: 1px solid #00000020;
line-height: 1.15em;
padding: 10px 5px;
}
.lightable-paper td {
vertical-align: middle;
border-bottom: 1px solid #00000010;
line-height: 1.15em;
padding: 7px 5px;
}
.lightable-paper.lightable-hover tbody tr:hover {
background-color: #F9EEC1;
}
.lightable-paper.lightable-striped tbody tr:nth-child(even) {
background-color: #00000008;
}
.lightable-paper.lightable-striped tbody td {
border: 0;
}
</style>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">MiRKAT Package Vignette</h1>
<h4 class="author">Anna Plantinga, Nehemiah Wilson, Haotian Zheng, Xiang
Zhan, Michael Wu, Jun Chen, Ni Zhao</h4>



<div id="overview" class="section level2">
<h2>Overview</h2>
<p>The MiRKAT package (v1.2.1) contains functions that test global
associations between the microbiota and different types of phenotypes,
such as univariate continuous or binary phenotypes, survival (censored
time-to-event) outcomes, longitudinal data, multivariate, and structured
phenotypes. For all these effects, the microbiome community effect is
modeled nonparametrically through a kernel function, which can
incorporate the phylogenetic tree information.</p>
<div id="changes-from-v1.1.0" class="section level3">
<h3>Changes from v1.1.0</h3>
<p>One additional function (MiRKAT.iQ), an integrated quantile
regression-based kernel association test, has been added in version
1.2.1.</p>
</div>
<div id="dependencies" class="section level3">
<h3>Dependencies</h3>
<p>The following packages are required for functions and examples in the
MiRKAT package: MASS, GUniFrac, CompQuadForm, quantreg, PearsonDS, lme4,
Matrix, permute, survival, and stats. All of these required packages are
available on CRAN.</p>
</div>
</div>
<div id="mirkat-microbiome-regression-based-kernel-association-test" class="section level2">
<h2>MiRKAT: Microbiome Regression-Based Kernel Association Test</h2>
<p>We begin by demonstrating the use of the Microbiome Regression-Based
Kernel Association Test (MiRKAT) for binary and continuous traits.</p>
<div id="example-dataset" class="section level3">
<h3>Example Dataset</h3>
<p>We use the throat microbiome data (Charlson et al 2010) from the
package GUniFrac to demonstrate our methods. Data are available for 60
subjects, of whom 28 were smokers and 32 were nonsmokers. Microbiome
data were collected from right and left nasopharynx and oropharynx
regions to form an OTU table with 856 OTUs. We want to evaluate whether
smoking is associated with differences in microbiome composition in the
upper respiratory tract, taking into consideration additional covariates
including gender and antibiotic use within 3 months. We also use
simulated data based on the throat dataset to demonstrate the usage of
MiRKAT-S, MMiRKAT, MiRKAT.R, MiRKAT.Q, and KRV.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;throat.otu.tab&quot;</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;throat.meta&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;throat.tree&quot;</span>)</span></code></pre></div>
</div>
<div id="data-preparation" class="section level3">
<h3>Data Preparation</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>Male <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(throat.meta<span class="sc">$</span>Sex <span class="sc">==</span> <span class="st">&quot;Male&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>Smoker <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(throat.meta<span class="sc">$</span>SmokingStatus <span class="sc">==</span> <span class="st">&quot;Smoker&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>anti <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(throat.meta<span class="sc">$</span>AntibioticUsePast3Months_TimeFromAntibioticUsage <span class="sc">!=</span> <span class="st">&quot;None&quot;</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>covar <span class="ot">&lt;-</span> <span class="fu">cbind</span>(Male, anti)</span></code></pre></div>
</div>
<div id="create-the-unifrac-distances" class="section level3">
<h3>Create the UniFrac Distances</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Note: may want to rarefy the data to the minimum sequencing depth for unweighted (presence/absence) distances </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="do">## otu.tab.rff &lt;- Rarefy(throat.otu.tab)$otu.tab.rff</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>unifracs <span class="ot">&lt;-</span> <span class="fu">GUniFrac</span>(throat.otu.tab, throat.tree, <span class="at">alpha =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>))<span class="sc">$</span>unifracs</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>D.weighted <span class="ot">=</span> unifracs[,,<span class="st">&quot;d_1&quot;</span>]</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>D.unweighted <span class="ot">=</span> unifracs[,,<span class="st">&quot;d_UW&quot;</span>]</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>D.generalized <span class="ot">=</span> unifracs[,,<span class="st">&quot;d_0.5&quot;</span>] </span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">requireNamespace</span>(<span class="st">&quot;vegan&quot;</span>)) {</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  D.BC <span class="ot">=</span> <span class="fu">as.matrix</span>(<span class="fu">vegdist</span>(throat.otu.tab, <span class="at">method=</span><span class="st">&quot;bray&quot;</span>))</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="convert-distance-to-kernel-matrices" class="section level3">
<h3>Convert Distance to Kernel Matrices</h3>
<p>The D2K function in MiRKAT converts distance matrices to kernel
matrices via the transformation <span class="math display">\[K =
-\frac{1}{2} \left(I - \frac{1 1&#39;}{n}\right) D^2 \left(I - \frac{1
1&#39;}{n}\right)\]</span> Here, <span class="math inline">\(I\)</span>
is the identity matrix and <span class="math inline">\(1\)</span> is an
<span class="math inline">\(n\)</span>-vector of ones. To ensure that
<span class="math inline">\(K\)</span> is positive semi-definite, we
replace negative eigenvalues with zero. That is, we perform an
eigenvalue decomposition <span class="math inline">\(K = U \Lambda
U\)</span>, where <span class="math inline">\(\Lambda =
\text{diag}(\lambda_1,...,\lambda_n)\)</span>, and then reconstruct the
kernel matrix using the non-negative eigenvalues <span class="math inline">\(\Lambda^* =
\text{diag}(\text{max}(\lambda_1,0),...,\text{max}(\lambda_n,0))\)</span>
so that <span class="math inline">\(K = U \Lambda^* U\)</span>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>K.weighted <span class="ot">=</span> <span class="fu">D2K</span>(D.weighted)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>K.unweighted <span class="ot">=</span> <span class="fu">D2K</span>(D.unweighted)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>K.generalized <span class="ot">=</span> <span class="fu">D2K</span>(D.generalized)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">requireNamespace</span>(<span class="st">&quot;vegan&quot;</span>)) {</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  K.BC <span class="ot">=</span> <span class="fu">D2K</span>(D.BC)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="testing-using-a-single-kernel" class="section level3">
<h3>Testing Using a Single Kernel</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MiRKAT</span>(<span class="at">y =</span> Smoker, <span class="at">X =</span> covar, <span class="at">Ks =</span> K.weighted, <span class="at">out_type =</span> <span class="st">&quot;D&quot;</span>, </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">method =</span> <span class="st">&quot;davies&quot;</span>, <span class="at">returnKRV =</span> <span class="cn">TRUE</span>, <span class="at">returnR2 =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## $p_values
## [1] 0.005306786
## 
## $KRV
## [1] 0.002884312
## 
## $R2
## [1] 0.01757917</code></pre>
<p>“Method” indicates which method the function should use to compute
the <strong>kernel-specific p-value</strong>. “davies” represents an
exact method that computes the p-value by inverting the characteristic
function of the mixture chi-square distribution. We adopt an exact
variance component test because most of the studies concerning
microbiome compositions have modest sample size. “permutation”
represents a residual permutation approach. “moment” represents an
approximation method that matches the first two moments. When
out_type=“C” (continuous outcome y), the “moment” method is the
Satterthwaite approximation. When out_type = “D” (dichotomous outcome),
the “moment” method is the small sample adjustment in Lee et al (2012).
When sample size is modest (<span class="math inline">\(n&lt;100\)</span> for continuous or <span class="math inline">\(n&lt;200\)</span> for dichotomous outcome), the
“moment” method can be inflated at very small size (such as <span class="math inline">\(\alpha\)</span> = 0.001), although the type I
error at <span class="math inline">\(\alpha\)</span> = 0.05 is usually
sustained. Therefore, we suggest using the Davies or permutation
approaches for such situations.</p>
</div>
<div id="measures-of-effect-size" class="section level3">
<h3>Measures of Effect Size</h3>
<p>If returnKRV = TRUE and/or returnR2 = TRUE, the function also returns
measures of effect size KRV and <span class="math inline">\(R^2\)</span>
(respectively). The KRV statistic is as described in However, we caution
that these measures of effect size remain very small (on the order of
1e-4 or smaller for the KRV statistic, and generally <span class="math inline">\(&lt;0.1\)</span> for R-squared) even when the
majority of the variability in the outcome is explainable by the
microbiome. Hence this should be used for internal comparisons rather
than external discussions of, e.g., percent variability explained.</p>
</div>
<div id="properties-of-different-kernels" class="section level3">
<h3>Properties of Different Kernels</h3>
<p>How to choose an appropriate distance matrix and kernel for testing
is a difficult question. However, it is important, since the distance
matrix used to generate the kernel strongly affects the power of our
tests. In particular, MiRKAT has highest power when the form of
association between the microbiota and the outcome assumed by the kernel
matches the true form of association. Poor choice of kernel will lead to
reduced power, although the type I error will be preserved.</p>
<p>In the case of the UniFrac families and the Bray-Curtis
dissimilarity, the factors at play are (1) the abundance of the
associated taxa and (2) whether closely related taxa (phylogenetically)
tend to be related or not related to the outcome as a group. For
example, the following are some of the distance metrics that have been
used for studies of the microbiome:</p>
<table>
<tbody>
<tr>
<td style="text-align:left;">
Distance
</td>
<td style="text-align:left;">
Phylogeny?
</td>
<td style="text-align:left;">
Abundance?
</td>
<td style="text-align:left;width: 20em; ">
Other notes
</td>
<td style="text-align:left;">
References
</td>
</tr>
<tr>
<td style="text-align:left;">
Unweighted UniFrac
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;">
No
</td>
<td style="text-align:left;width: 20em; ">
</td>
<td style="text-align:left;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
Weighted UniFrac
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;width: 20em; ">
</td>
<td style="text-align:left;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Generalized UniFrac
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;">
(Yes)
</td>
<td style="text-align:left;width: 20em; ">
Parameter alpha defines extent to which abundance is taken into account
</td>
<td style="text-align:left;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Jaccard
</td>
<td style="text-align:left;">
No
</td>
<td style="text-align:left;">
No
</td>
<td style="text-align:left;width: 20em; ">
1 - (taxa in both)/(taxa in either); typically presence/absence, but can
be extended to an abundance-weighted version
</td>
<td style="text-align:left;">
4,5
</td>
</tr>
<tr>
<td style="text-align:left;">
Bray-Curtis
</td>
<td style="text-align:left;">
No
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;width: 20em; ">
Similar to Jaccard, but uses counts
</td>
<td style="text-align:left;">
6
</td>
</tr>
</tbody>
</table>
<p>In the table above, “Yes” indicates the distance or dissimilarity
metric has the feature; “(Yes)” indicates that the feature is present
either in some variations of the metric or is present to some extent;
and “No” indicates that the feature is not present.</p>
<p>Depending on which of these characteristics are expected to be
present in a particular study (based on prior knowledge or intuition),
an appropriate distance or dissimilarity can be selected. If the study
is exploratory and strong protection of type 1 error is not needed,
several distance metrics can be explored. Depending on which one(s) are
highly significant, some information can be gained about the nature of
any association between the microbiota and the outcome.</p>
</div>
<div id="testing-using-multiple-kernels" class="section level3">
<h3>Testing Using Multiple Kernels</h3>
<p>We provide an omnibus test that takes into account multiple kernels
simultaneously. The method is robust in the sense that it has
substantial power gain compared to when an improper kernel is used, and
has little power loss compared to when the best kernel is used. Two
options for omnibus testing are provided: residual permutation and the
Cauchy combination test (weighting all kernels equally).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">requireNamespace</span>(<span class="st">&quot;vegan&quot;</span>)) {</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  Ks <span class="ot">=</span> <span class="fu">list</span>(<span class="at">K.weighted =</span> K.weighted, <span class="at">K.unweighted =</span> K.unweighted, <span class="at">K.BC =</span> K.BC)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  Ks <span class="ot">=</span> <span class="fu">list</span>(<span class="at">K.weighted =</span> K.weighted, <span class="at">K.unweighted =</span> K.unweighted)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="fu">MiRKAT</span>(<span class="at">y =</span> Smoker, <span class="at">Ks =</span> Ks, <span class="at">X =</span> covar, <span class="at">out_type =</span> <span class="st">&quot;D&quot;</span>, </span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">method =</span> <span class="st">&quot;davies&quot;</span>, <span class="at">omnibus =</span> <span class="st">&quot;permutation&quot;</span>, </span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">returnKRV =</span> <span class="cn">TRUE</span>, <span class="at">returnR2 =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##  0.005306786  0.003801015  0.002287489 
## 
## $omnibus_p
## [1] 0.011
## 
## $KRV
##   K.weighted K.unweighted         K.BC 
## 0.0028843118 0.0010877663 0.0006235506 
## 
## $R2
##   K.weighted K.unweighted         K.BC 
##   0.01757917   0.01981955   0.02403491</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MiRKAT</span>(<span class="at">y =</span> Smoker, <span class="at">Ks =</span> Ks, <span class="at">X =</span> covar, <span class="at">out_type =</span> <span class="st">&quot;D&quot;</span>, </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">method =</span> <span class="st">&quot;davies&quot;</span>, <span class="at">omnibus =</span> <span class="st">&quot;cauchy&quot;</span>, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">returnKRV =</span> <span class="cn">FALSE</span>, <span class="at">returnR2 =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##  0.005306786  0.003801015  0.002287489 
## 
## $omnibus_p
## [1] 0.003375785</code></pre>
</div>
<div id="testing-with-a-continuous-outcome-variable" class="section level3">
<h3>Testing with a Continuous Outcome Variable</h3>
<p>The only difference between using a continuous and dichotomous
outcome variable is the value of the argument “out_type”, as shown
below.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="fu">nrow</span>(K.weighted))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">MiRKAT</span>(<span class="at">y =</span> y, <span class="at">Ks =</span> Ks, <span class="at">X =</span> covar, <span class="at">out_type =</span> <span class="st">&quot;C&quot;</span>, <span class="at">nperm =</span> <span class="dv">999</span>, </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">method =</span> <span class="st">&quot;davies&quot;</span>, <span class="at">omnibus =</span> <span class="st">&quot;cauchy&quot;</span>, <span class="at">returnKRV =</span> <span class="cn">TRUE</span>, <span class="at">returnR2 =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##    0.6578367    0.5912145    0.8011834 
## 
## $omnibus_p
## [1] 0.702978
## 
## $KRV
##   K.weighted K.unweighted         K.BC 
## 0.0007495959 0.0006680155 0.0001768607 
## 
## $R2
##   K.weighted K.unweighted         K.BC 
##  0.001187321  0.007474732  0.001933579</code></pre>
<p>This function outputs p-values for association using each single
kernel and an omnibus p-value considering all kernels. The omnibus
p-value is obtained through residual permutation where the minimum
p-values from each of the individual tests are used as test statistics.
As before, the “method” option only indicates the method that is used
for generating individual kernel p-values.</p>
</div>
</div>
<div id="mirkat-s-survival-outcome" class="section level2">
<h2>MiRKAT-S: Survival Outcome</h2>
<div id="simulate-time-to-event-data" class="section level3">
<h3>Simulate time to event data</h3>
<p>We again use Charlson’s throat microbiome data to demonstrate the use
of MiRKAT-S. Data loading and preparation are the same as in the
previous section. Because the original dataset has a binary phenotype
(smoking) rather than a measure of censored time to event outcomes, we
consider smoking status and gender as covariates and generate null
outcome data from the Exponential distribution. Specifically, we
generate survival times as <span class="math inline">\(S \sim
\text{Exponential}(1 + I(\text{smoke}) + I(\text{male}))\)</span>, and
censoring times as <span class="math inline">\(C \sim
\text{Exponential}(0.75)\)</span>. Then the observed outcome measures
are observation time <span class="math inline">\(T = \text{min}(S,
C)\)</span> and an indicator variable for whether the event was
observed, <span class="math inline">\(\Delta = I(S \leq C)\)</span>.
That is, when delta = 1, the corresponding “obstime” is the survival
time, and when delta = 0, the corresponding observation is censored and
“obstime” is the time of censoring. This simulation procedure results in
approximately 33% censoring.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate outcomes</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Here, outcome is associated with covariates but unassociated with microbiota</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Approximately 33% censoring</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>SurvTime <span class="ot">&lt;-</span> <span class="fu">rexp</span>(<span class="dv">60</span>, (<span class="dv">1</span> <span class="sc">+</span> Smoker <span class="sc">+</span> Male))</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>CensTime <span class="ot">&lt;-</span> <span class="fu">rexp</span>(<span class="dv">60</span>, <span class="fl">0.75</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>Delta <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(SurvTime <span class="sc">&lt;=</span> CensTime )</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>ObsTime <span class="ot">&lt;-</span> <span class="fu">pmin</span>(SurvTime, CensTime)</span></code></pre></div>
<p>The p-value for the test may be generated using permutation or
Davies’ exact method. Davies’ exact method, which computes the p-value
based on a mixture of chi-square distributions, is used when “perm = F”.
We use a small-sample correction to account for the modest sample sizes
and sparse OTU count matrices that often result from studies of the
microbiome.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Davies&#39; exact method </span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">MiRKATS</span>(<span class="at">obstime =</span> ObsTime, <span class="at">delta =</span> Delta, <span class="at">X =</span> <span class="fu">cbind</span>(Smoker, Male, anti), <span class="at">Ks =</span> Ks, </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">perm =</span> F, <span class="at">omnibus =</span> <span class="st">&quot;cauchy&quot;</span>, <span class="at">returnKRV =</span> T, <span class="at">returnR2 =</span> T)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##    0.7659192    0.5528894    0.9597768 
## 
## $omnibus_p
## [1] 0.8990951
## 
## $KRV
##   K.weighted K.unweighted         K.BC 
## 0.0007273783 0.0005514495 0.0001271282 
## 
## $R2
##   K.weighted K.unweighted         K.BC 
##  0.001117981  0.005093709  0.000999040</code></pre>
<p>Using “perm = T” indicates that a permutation p-value should be
calculated for each kernel-specific test. Overall, permutation is
recommended when the sample size is small, as Davies’ method may be
slightly anti-conservative with very small sample sizes. MiRKAT-S will
generate a warning when permutation is not used for sample sizes <span class="math inline">\(n \leq 50\)</span>. “nperm” indicates the number
of permutations to perform to generate the p-value (default = 1000).</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Permutation </span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">MiRKATS</span>(<span class="at">obstime =</span> ObsTime, <span class="at">delta =</span> Delta, <span class="at">X =</span> <span class="fu">cbind</span>(Smoker, Male, anti), <span class="at">Ks =</span> Ks, </span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">perm =</span> T, <span class="at">omnibus =</span> <span class="st">&quot;cauchy&quot;</span>, <span class="at">returnKRV =</span> T, <span class="at">returnR2 =</span> T)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##    0.7807808    0.5665666    0.9679680 
## 
## $omnibus_p
## [1] 0.9176127
## 
## $KRV
##   K.weighted K.unweighted         K.BC 
## 0.0007273783 0.0005514495 0.0001271282 
## 
## $R2
##   K.weighted K.unweighted         K.BC 
##  0.001117981  0.005093709  0.000999040</code></pre>
<p>As above, the omnibus p-value may be generated either using residual
permutation to create a null distribution of minimum p-values, and the
minimum p-value from the original MiRKAT-S analysis is tested against
this distribution as the test statistic of the omnibus test, or using
the Cauchy combination test. The permutation-based omnibus test is
described further at <a href="https://github.com/hk1785/OMiSA" class="uri">https://github.com/hk1785/OMiSA</a> (Koh 2018, DOI: <a href="https://doi.org/10.1186/s12864-018-4599-8" class="uri">https://doi.org/10.1186/s12864-018-4599-8</a>).</p>
</div>
</div>
<div id="mmirkat-multivariate-continuous-outcome" class="section level2">
<h2>MMiRKAT: Multivariate Continuous Outcome</h2>
<p>MMiRKAT is designed to test the association between the microbiome
and a low-dimensional multivariate continuous outcome. For structured or
high dimensional outcomes, the kernel RV test (KRV) is recommended
instead, as demonstrated in the next section.</p>
<p>We again use Charlson’s throat microbiome data to demonstrate the use
of MMiRKAT. We generate multivariate outcomes <span class="math inline">\(Y \sim N_{3n}(0, I_{3n})\)</span>.</p>
<p>MMiRKAT uses a small-sample correction procedure for p-value
calculation. There is no omnibus test in MMiRKAT; only kernel-specific
p-values and, optionally, the KRV and R-squared statistics are
returned.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">nrow</span>(throat.otu.tab)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>Y <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(n<span class="sc">*</span><span class="dv">3</span>, <span class="dv">0</span>, <span class="dv">1</span>), n, <span class="dv">3</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">MMiRKAT</span>(<span class="at">Y =</span> Y, <span class="at">X =</span> <span class="fu">cbind</span>(Male, anti), <span class="at">Ks =</span> Ks, </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">returnKRV =</span> <span class="cn">TRUE</span>, <span class="at">returnR2 =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##    0.1073553    0.2126433    0.2922759 
## 
## $KRV
##   K.weighted K.unweighted         K.BC 
## 0.0012759596 0.0007062962 0.0002528821 
## 
## $R2
##   K.weighted K.unweighted         K.BC 
##   0.01036837   0.02518368   0.01191402</code></pre>
</div>
<div id="krv-high-dimensional-structured-outcome" class="section level2">
<h2>KRV: High-Dimensional Structured Outcome</h2>
<p>KRV (kernel RV coefficient) is designed to evaluate the association
between microbiome composition and a structured, potentially
high-dimensional phenotype, such as gene expression of a set of genes
which are functionally related. The KRV statistic can capture nonlinear
associations and complex relationships within the individual data types
and between the complex multivariate phenotype and microbiome
composition by measuring general dependency.</p>
<p>Two kernels are involved in KRV statistics: one kernel for microbiome
composition, which can be obtained by transforming the distance metrics
as in the previous sections, and one kernel capturing similarities in
the phenotypes. As an alternative, KRV can also take as input a kernel
matrix for microbiome composition, a multivariate phenotype y, and a set
of additional covariates X. In this case, a multivariate regression is
carried out, and then the residuals of this regression are subsequently
used to construct a Gaussian or linear kernel matrix.</p>
<p>An omnibus test has also been implemented for the KRV function. This
omnibus test is different from the other omnibus tests in the package,
as it creates an “omnibus kernel matrix” and then performs the KRV
analysis on that omnibus kernel (option: omnibus = “kernel_om”). The
omnibus kernel matrix is constructed via the second formulation in Zhan
et al (2017), an unsupervised weighted combination of kernels that the
authors found to have the best performance across a range of simulation
settings. Specifically, the omnibus kernel is defined as <span class="math display">\[K_{om} = \sum_{i=1}^m
\frac{K_i}{tr(K_i)}.\]</span> The p-value from this single omnibus
kernel is returned as the omnibus p-value for KRV. Alternatively, the
Cauchy combination test may be used (option: omnibus = “Cauchy”).</p>
<div id="create-the-kernel-matrix-of-simulated-data" class="section level3">
<h3>Create the Kernel Matrix of Simulated Data</h3>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>rho <span class="ot">=</span> <span class="fl">0.2</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>Va <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">rep</span>(rho, (<span class="dv">2</span><span class="sc">*</span>n)<span class="sc">^</span><span class="dv">2</span>), <span class="dv">2</span><span class="sc">*</span>n, <span class="dv">2</span><span class="sc">*</span>n)<span class="sc">+</span><span class="fu">diag</span>(<span class="dv">1</span><span class="sc">-</span>rho, <span class="dv">2</span><span class="sc">*</span>n)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>G <span class="ot">=</span> <span class="fu">mvrnorm</span>(n, <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">2</span><span class="sc">*</span>n), Va)</span></code></pre></div>
</div>
<div id="testing-using-an-outcome-kernel-matrix" class="section level3">
<h3>Testing Using an Outcome Kernel Matrix</h3>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>lin.K.y <span class="ot">=</span> G <span class="sc">%*%</span> <span class="fu">t</span>(G)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Unadjusted</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="fu">KRV</span>(<span class="at">kernels.otu =</span> Ks, <span class="at">kernel.y =</span> lin.K.y, </span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">omnibus =</span> <span class="st">&quot;kernel_om&quot;</span>, <span class="at">adjust.type =</span> <span class="cn">NULL</span>, </span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">returnKRV =</span> <span class="cn">TRUE</span>, <span class="at">returnR2 =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##    0.6231597    0.4956971    0.6321714 
## 
## $omnibus_p
## [1] 0.6134599
## 
## $KRV
##   K.weighted K.unweighted         K.BC 
## 1.713514e-04 1.226831e-04 4.210276e-05 
## 
## $R2
##   K.weighted K.unweighted         K.BC 
##   0.03958771   0.16086562   0.06991840</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Adjusting both </span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">KRV</span>(<span class="at">kernels.otu =</span> Ks, <span class="at">kernel.y =</span> lin.K.y, </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> <span class="fu">cbind</span>(Male, anti), </span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">omnibus =</span> <span class="st">&quot;kernel_om&quot;</span>, <span class="at">adjust.type =</span> <span class="st">&quot;both&quot;</span>, </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">returnKRV =</span> <span class="cn">TRUE</span>, <span class="at">returnR2 =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##    0.3005123    0.1611390    0.2750115 
## 
## $omnibus_p
## [1] 0.2474459
## 
## $KRV
##   K.weighted K.unweighted         K.BC 
## 2.005005e-04 1.345159e-04 4.849817e-05 
## 
## $R2
##   K.weighted K.unweighted         K.BC 
##   0.04568613   0.17001701   0.07888428</code></pre>
</div>
<div id="testing-with-multiple-kernels-using-the-covariate-option" class="section level3">
<h3>Testing with Multiple Kernels, Using the Covariate Option</h3>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Adjust both kernel matrices</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">KRV</span>(<span class="at">y =</span> G, <span class="at">X =</span> <span class="fu">cbind</span>(Male, anti), <span class="at">adjust.type =</span> <span class="st">&quot;both&quot;</span>, </span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">kernels.otu =</span> Ks, <span class="at">kernel.y =</span> <span class="st">&quot;linear&quot;</span>)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##    0.3005123    0.1611390    0.2750115 
## 
## $omnibus_p
## [1] 0.2474459</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Adjust phenotype only</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">KRV</span>(<span class="at">y =</span> G, <span class="at">X =</span> <span class="fu">cbind</span>(Male, anti), <span class="at">adjust.type =</span> <span class="st">&quot;phenotype&quot;</span>, </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">kernels.otu =</span> Ks, <span class="at">kernel.y =</span> <span class="st">&quot;Gaussian&quot;</span>)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##    0.8039612    0.7509481    0.8588011 
## 
## $omnibus_p
## [1] 0.8403911</code></pre>
</div>
</div>
<div id="robust-mirkat" class="section level2">
<h2>Robust MiRKAT</h2>
<p>MiRKAT is sensitive to strong skewness or outliers in continuous
outcome variables, but many biological features have non-Gaussian
distributions. To accommodate such features, rather than using OLS,
robust MiRKAT uses a robust regression method. Specifically, a kernel
matrix for the outcome is created using a linear kernel based on the
residuals from robust linear regression (MiRKAT-R) or quantile
regression (MiRKAT-Q) of the outcome on non-microbiome covariates, then
related to the microbiome kernel using the KRV test. Note that if there
are no additional covariates, MiRKAT-R and MiRKAT-Q simplify to
MiRKAT.</p>
<p>We additionally note that quantile regression has lower power than
robust regression in every scenario considered, and hence recommend use
of MiRKAT-R in most circumstances. Both functions call the KRV function
once an outcome kernel matrix has been created using quantile or robust
regression, so the omnibus tests for MiRKAT.Q and MiRKAT.R are the same
as for KRV.</p>
<div id="mirkat.r-continuous-outcomes-with-robust-linear-regression" class="section level3">
<h3>MiRKAT.R: Continuous Outcomes with Robust Linear Regression</h3>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">rchisq</span>(<span class="fu">nrow</span>(K.weighted), <span class="dv">1</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">MiRKAT.R</span>(<span class="at">y =</span> y, <span class="at">X =</span> <span class="fu">cbind</span>(Male, anti), <span class="at">Ks =</span> Ks)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##    0.5502230    0.9116325    0.3288858 
## 
## $omnibus_p
## [1] 0.5506275</code></pre>
</div>
<div id="mirkat.q-continuous-outcomes-with-quantile-regression" class="section level3">
<h3>MiRKAT.Q: Continuous Outcomes with Quantile Regression</h3>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MiRKAT.Q</span>(<span class="at">y =</span> y, <span class="at">X=</span> <span class="fu">cbind</span>(Male, anti), <span class="at">Ks =</span> Ks)</span></code></pre></div>
<pre><code>## $p_values
##   K.weighted K.unweighted         K.BC 
##    0.6362740    0.4675527    0.6828106 
## 
## $omnibus_p
## [1] 0.6336712</code></pre>
</div>
</div>
<div id="glmm-mirkat-longitudinalcorrelated-continuous-binary-or-count-data-generalized-linear-mixed-model" class="section level2">
<h2>GLMM-MiRKAT: Longitudinal/Correlated Continuous, Binary, or Count
Data (Generalized Linear Mixed Model)</h2>
<p>GLMM-MiRKAT allows for association testing between the microbiome and
Guassian, binomial (e.g. disease status, treatment/placebo) and Poisson
(e.g. number of tumors/treatments) traits with correlated data resulting
from longitudinal studies or family-based studies. For Gaussian traits,
calculation of P-values may use the Davies method and Cauchy combination
test or permutation; other outcome types only permit a permuation
approach for p-value and omnibus p-value calculations.</p>
<p>We switch to a different simulation approach for these last two
functions, as they both deal with correlated data.</p>
<div id="prepare-the-data" class="section level3">
<h3>Prepare the Data</h3>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import example microbiome data with Gaussian traits</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(nordata)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract microbiome and meta information</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>otu.tab <span class="ot">&lt;-</span> nordata<span class="sc">$</span>nor.otu.tab</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> nordata<span class="sc">$</span>nor.tree</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> nordata<span class="sc">$</span>nor.meta</span></code></pre></div>
</div>
<div id="create-the-unifrac-distances-1" class="section level3">
<h3>Create the UniFrac Distances</h3>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>Ds <span class="ot">&lt;-</span> <span class="fu">GUniFrac</span>(otu.tab, tree, <span class="at">alpha=</span><span class="fu">c</span>(<span class="fl">0.5</span>, <span class="dv">1</span>))<span class="sc">$</span>unifracs</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>D.unweighted <span class="ot">&lt;-</span> Ds[,,<span class="st">&quot;d_UW&quot;</span>]</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>D.generalized <span class="ot">&lt;-</span> Ds[,,<span class="st">&quot;d_0.5&quot;</span>]</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>D.weighted <span class="ot">&lt;-</span> Ds[,,<span class="st">&quot;d_1&quot;</span>]</span></code></pre></div>
</div>
<div id="convert-distance-to-kernel-matrices-1" class="section level3">
<h3>Convert Distance to Kernel Matrices</h3>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>K.unweighted <span class="ot">&lt;-</span> <span class="fu">D2K</span>(D.unweighted)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>K.generalized <span class="ot">&lt;-</span> <span class="fu">D2K</span>(D.generalized)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>K.weighted <span class="ot">&lt;-</span> <span class="fu">D2K</span>(D.weighted)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>Ks <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">K.weighted =</span> K.weighted, </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">K.unweighted =</span> K.unweighted, </span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>           <span class="at">K.generalized =</span> K.generalized)</span></code></pre></div>
</div>
<div id="example-1.-gaussian-traits-e.g.-body-mass-index" class="section level3">
<h3>Example 1. Gaussian traits (e.g., body mass index)</h3>
<p>GLMM-MiRKAT returns kernel-specific and omnibus p-values. No measures
of effect size are available for GLMM-MiRKAT.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># GLMM-MiRKAT with computational p-values (Davies + Cauchy combination test)</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">GLMMMiRKAT</span>(<span class="at">y =</span> meta<span class="sc">$</span>y, <span class="at">X =</span> <span class="fu">cbind</span>(meta<span class="sc">$</span>x1, meta<span class="sc">$</span>x2), <span class="at">id =</span> meta<span class="sc">$</span>id, <span class="at">Ks =</span> Ks, </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">model =</span> <span class="st">&quot;gaussian&quot;</span>, <span class="at">method =</span> <span class="st">&quot;davies&quot;</span>, <span class="at">omnibus =</span> <span class="st">&quot;Cauchy&quot;</span>)</span></code></pre></div>
<pre><code>## $p_values
##    K.weighted  K.unweighted K.generalized 
##  0.0001581617  0.5012159556  0.0236046037 
## 
## $omnibus_p
## [1] 0.0004713334</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># GLMM-MiRKAT with permutation test</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">GLMMMiRKAT</span>(<span class="at">y =</span> meta<span class="sc">$</span>y, <span class="at">X =</span> <span class="fu">cbind</span>(meta<span class="sc">$</span>x1, meta<span class="sc">$</span>x2), <span class="at">id =</span> meta<span class="sc">$</span>id, <span class="at">Ks =</span> Ks, </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">model =</span> <span class="st">&quot;gaussian&quot;</span>)</span></code></pre></div>
<pre><code>## $p_values
##    K.weighted  K.unweighted K.generalized 
##    0.00039992    0.54149170    0.07018596 
## 
## $omnibus_p
## [1] 0.00079984</code></pre>
</div>
<div id="example-2.-binomial-traits-e.g.-disease-status-treatmentplacebo" class="section level3">
<h3>Example 2. Binomial traits (e.g., disease status,
treatment/placebo)</h3>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a> <span class="co"># Import microbiome data with binomial traits</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a> <span class="fu">data</span>(bindata)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract microbiome and meta information</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="co"># (Microbiome is the same as for Gaussian outcomes)</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>otu.tab <span class="ot">&lt;-</span> bindata<span class="sc">$</span>bin.otu.tab</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> bindata<span class="sc">$</span>bin.tree</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> bindata<span class="sc">$</span>bin.meta</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Run GLMM-MiRKAT</span></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a><span class="fu">GLMMMiRKAT</span>(meta<span class="sc">$</span>y, <span class="at">X =</span> <span class="fu">cbind</span>(meta<span class="sc">$</span>x1, meta<span class="sc">$</span>x2), <span class="at">id =</span> meta<span class="sc">$</span>id, </span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>           <span class="at">Ks =</span> Ks, <span class="at">model =</span> <span class="st">&quot;binomial&quot;</span>)</span></code></pre></div>
<pre><code>## $p_values
##    K.weighted  K.unweighted K.generalized 
##     0.9198160     0.6312737     0.8388322 
## 
## $omnibus_p
## [1] 0.8112378</code></pre>
</div>
<div id="example-3.-poisson-traits-e.g.-number-of-tumors" class="section level3">
<h3>Example 3. Poisson traits (e.g., number of tumors)</h3>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import microbiome data with Poisson traits</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(poisdata)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract microbiome and meta information</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="co"># (Microbiome is again the same)</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>otu.tab <span class="ot">&lt;-</span> poisdata<span class="sc">$</span>pois.otu.tab</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> poisdata<span class="sc">$</span>pois.tree</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> poisdata<span class="sc">$</span>pois.meta</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Run GLMM-MiRKAT</span></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a><span class="fu">GLMMMiRKAT</span>(meta<span class="sc">$</span>y, <span class="at">X =</span> <span class="fu">cbind</span>(meta<span class="sc">$</span>x1, meta<span class="sc">$</span>x2), <span class="at">id =</span> meta<span class="sc">$</span>id, </span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>           <span class="at">Ks =</span> Ks, <span class="at">model =</span> <span class="st">&quot;poisson&quot;</span>)</span></code></pre></div>
<pre><code>## $p_values
##    K.weighted  K.unweighted K.generalized 
##     0.1453709     0.1753649     0.2485503 
## 
## $omnibus_p
## [1] 0.2841432</code></pre>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
